crmsh/ui_node.py

```diff
@@ -287,36 +290,87 @@ def do_print(uname):
-        'usage: standby [<node>] [<lifetime>]'
-        argl = list(args)
-        node = None
-        lifetime = utils.fetch_lifetime_opt(argl, iso8601=False)
-        if not argl:
-            node = utils.this_node()
-        elif len(argl) == 1:
-            node = args[0]
-            if not xmlutil.is_our_node(node):
-                logger.error("%s: node name not recognized", node)
-                return False
-        else:
-            logger_utils.syntax_err(args, context=context.get_command_name())
-            return False
-        opts = ''
+        """
+        usage: standby [<node>] [<lifetime>]
+        To avoid race condition for --all option, melt all standby values into one cib replace session
+        """
+        # Parse lifetime option
+        lifetime_opt = "forever"
+        lifetime = utils.fetch_lifetime_opt(list(args), iso8601=False)
         if lifetime:
-            opts = "--lifetime='%s'" % lifetime
-        else:
-            opts = "--lifetime='forever'"
-        return utils.ext_cmd(self.node_standby % (node, "on", opts)) == 0
+            lifetime_opt = lifetime
+            args = args[:-1]
+
+        # Parse node option
+        node_list = parse_option_for_nodes(context, *args)
+        for node in node_list[:]:
+            if utils.is_standby(node):
+                logger.info("Node %s already standby", node)
+                node_list.remove(node)
+        if not node_list:
+            return
+
+        # For default "forever" lifetime, under "nodes" section
+        xml_path = constants.XML_NODE_PATH
+        xml_query_path = constants.XML_NODE_QUERY_STANDBY_PATH
+        standby_template = constants.STANDBY_TEMPLATE
+        # For "reboot" lifetime, under "status" section
+        if lifetime_opt == "reboot":
+            xml_path = constants.XML_STATUS_PATH
+            xml_query_path = constants.XML_STATUS_QUERY_STANDBY_PATH
+            standby_template = constants.STANDBY_TEMPLATE_REBOOT
+
+        cib = xmlutil.cibdump2elem()
+        xml_item_list = cib.xpath(xml_path)
+        for xml_item in xml_item_list:
+            if xml_item.get("uname") in node_list:
+                node_id = xml_item.get('id')
+                # If the standby nvpair already exists, continue
+                if cib.xpath(xml_query_path.format(node_id=node_id)):
+                    continue
+                # Create standby nvpair
+                standby_template_str = standby_template.format(node_id=node_id, value="on")
+                xml_item.append(etree.fromstring(standby_template_str))
+
+        cib_str = xmlutil.xml_tostring(cib)
+        # Consider both "nodes" and "status" section might contain different standby value at the same time
+        # Should replace all possible "off" values here
+        for node in node_list:
+            node_id = utils.get_nodeid_from_name(node)
+            cib_str = re.sub(constants.STANDBY_NV_RE.format(node_id=node_id, value="off"), r'\1value="on"\2', cib_str)
```

Comment:
It doesn't seem to cover the cases where it's set with other valid boolean values "false/no/n/0"?  How about substituting that with standby=on anyway?

And technically, the id of such an nvpair doesn't have to contain the node_id. Could we rely on the results of xpath queries like done in above and do the modifications?

Reply:
Good point! Thanks!

Consider need more time to verify these corner cases, can I put these into my backlog and improve later?

Reply:
Sure. We can improve that later.

Reply:
PR to improve this: https://github.com/ClusterLabs/crmsh/pull/893